aten/src/ATen/native/mkldnn/xpu/detail/QInt4Matmul.cpp

```diff
@@ -0,0 +1,671 @@
+        /* mask */ 0,
+        {},
+        memory::data_type::s8);
+  } else {
+    pattr.set_zero_points(
+        DNNL_ARG_WEIGHTS,
+        /* mask */ (1 << 0) + (1 << 1),
+        {group_size, 1},
+        memory::data_type::u4);
+  }
+  pattr.set_fpmath_mode(dnnl::fpmath_mode::f16, true);
+}
+
+template <typename F>
+static at::Tensor dnnl_matmul_w4a16_common(
+    Tensor& result, // dst, [b, m, n]
+    const Tensor& mat1_, // src, [b, m, k]
+    const Tensor& mat2, // quantized weight, [K/8, N] transpose
+    const c10::optional<Tensor>& bias,
+    const Tensor& scale, // [k/group_size, n]
+    const Tensor& zp, // [k/group_size, n/8]
+    int64_t group_size,
+    bool m2_trans,
+    const c10::optional<Tensor>& g_idx,
+    F pattr,
+    Tensor res_flat,
+    Tensor res1_flat) {
+  // For GPTQ with desc_act=True scenario
+  auto mat1 = g_idx.has_value() ? mat1_.index_select(-1, g_idx.value()) : mat1_;
+
+  auto src_sz = mat1.sizes();
+  auto o_sz = mat1.sizes().vec();
+  auto b_sz = mat2.sizes();
+  *(o_sz.end() - 1) = *(b_sz.end() - 1);
+  result = at::empty(o_sz, mat1.options());
+
+  const int m = std::reduce(
+      src_sz.begin(), src_sz.end() - 1, 1, std::multiplies<int64_t>());
+  const int n = b_sz[1]; // presume channel last format
+  const int k = *(src_sz.end() - 1);
+
+  // get device, engine, stream
+  const int device_id = c10::xpu::current_device();
+  at::Device curDevice = at::Device(at::kXPU, device_id);
+  auto engine = GpuEngineManager::Instance().get_engine(curDevice);
+
+  joint_dtypes jd;
+  if (mat1.scalar_type() == at::ScalarType::Half) {
+    jd = joint_dtypes::_f16_int4;
+  } else if (mat1.scalar_type() == at::ScalarType::BFloat16) {
+    jd = joint_dtypes::_bf16_int4;
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        false, "Unsupported data type for int4 matmul: ", mat1.scalar_type());
+  }
+
+  // TODO: add bias datatype according to onednn requirement
+  bias_type_t b_type;
+  if (bias.has_value()) {
+    auto& b = bias.value();
+    const auto nuelm = b.numel();
+    if (nuelm == 1) {
+      b_type = bias_type_t::_scalar;
+    } else if (nuelm == m * n) {
+      b_type = bias_type_t::_mn;
+    } else if (b.size(b.dim() - 1) == n && nuelm == n) {
+      b_type = bias_type_t::_n;
+    } else if (b.size(b.dim() - 1) == 1 && nuelm == m) {
+      b_type = bias_type_t::_m;
+    } else {
+      TORCH_CHECK(0, "unsupported bias dim in matmul ...", b.sizes());
+    }
+  } else {
+    b_type = bias_type_t::_none;
+  }
+
+  const int64_t ldb = mat2.strides()[mat2.dim() - 1] * 8; // for int4 matmul
+  const int64_t lda = mat1.strides()[mat1.dim() - 2];
+  const int64_t ldc = result.strides()[result.dim() - 2];
+
+  trans_type tt = trans_type::_nt; // only support nt for int4 matmul
+  auto& matmul_ext = matmul_primitive_create_and_cache(
+      jd, tt, b_type, m, n, k, lda, ldb, ldc, device_id, pattr);
+
+  int arg_off = 0;
+  // set scale and zero point for matmul args
+  matmul_ext.set_attribute(
+      arg_off++,
```

Comment:
using unordered_map in `primitive_ext` would help hiding this implementation detai.